File: src/main/java/com/leetcode/datastructures/trees/MergeTwoBinaryTrees.java
package com.leetcode.datastructures.trees;

/**
 * https://leetcode.com/problems/merge-two-binary-trees/#/description
 */
public class MergeTwoBinaryTrees {

    public TreeNode mergeTrees(TreeNode t1, TreeNode t2) {
        return traverse(t1, t2);
    }

    private TreeNode traverse(TreeNode t1, TreeNode t2) {
        if (t1 != null && t2 != null) {
            TreeNode curr = new TreeNode(t1.val + t2.val);
            curr.left = traverse(t1.left, t2.left);
            curr.right = traverse(t1.right, t2.right);

            return curr;
        }

        if (t1 != t2) {
            return t1 == null ? t2 : t1;
        } else {
            return null;
        }
    }

}

File: src/main/kotlin/com/leetcode/trees/MergeTwoBinaryTrees.kt
package com.leetcode.trees

class TreeNode(var `val`: Int = 0) {
    var left: TreeNode? = null
    var right: TreeNode? = null
}

fun mergeTrees(t1: TreeNode?, t2: TreeNode?): TreeNode? {
    if (t1 == null && t2 == null) return null

    val node = merge(t1, t2)

    node?.left = mergeTrees(t1?.left, t2?.left)
    node?.right = mergeTrees(t1?.right, t2?.right)

    return node
}

private fun merge(t1: TreeNode?, t2: TreeNode?): TreeNode? {
    if (t1 == null && t2 == null) return null

    return TreeNode((t1?.`val` ?: 0) + (t2?.`val` ?: 0))
}